clients/stellar-relay-lib/src/tests/mod.rs

```diff
@@ -84,41 +97,54 @@ async fn stellar_overlay_should_receive_tx_set() {
 
 	let node_info = NodeInfo::new(19, 25, 23, "v19.5.0".to_string(), &PUBLIC_NETWORK);
 	let cfg = ConnConfig::new(TIER_1_VALIDATOR_IP_PUBLIC, 11625, secret, 0, true, true, false);
-	//act
-	let mut overlay_connection =
-		StellarOverlayConnection::connect(node_info.clone(), cfg).await.unwrap();
 
-	let mut tx_set_vec = vec![];
-	let mut attempt = 0;
-	while let Some(relay_message) = overlay_connection.listen().await {
-		if attempt > 300 {
-			break
-		}
-		attempt += 1;
-		match relay_message {
-			StellarRelayMessage::Data { p_id: _, msg_type: _, msg } => match *msg {
-				StellarMessage::ScpMessage(msg) => {
-					if let ScpStatementPledges::ScpStExternalize(stmt) = &msg.statement.pledges {
-						let txset_hash = get_tx_set_hash(stmt);
-						overlay_connection
-							.send(StellarMessage::GetTxSet(txset_hash))
-							.await
-							.unwrap();
-					}
-				},
-				StellarMessage::TxSet(set) => {
-					tx_set_vec.push(set);
-					break
+	let mut overlay_connection = Arc::new(Mutex::new(
+		StellarOverlayConnection::connect(node_info.clone(), cfg).await.unwrap(),
+	));
+
+	let mut ov_conn = overlay_connection.clone();
+	let tx_set_vec = Arc::new(Mutex::new(vec![]));
+
+	let mut tx_set_vec_clone = tx_set_vec.clone();
+
+	timeout(Duration::from_secs(300), async move {
+		let mut ov_conn_locked = ov_conn.lock().await;
+
+		while let Some(relay_message) = ov_conn_locked.listen().await {
+			match relay_message {
+				StellarRelayMessage::Data { p_id: _, msg_type: _, msg } => match *msg {
+					StellarMessage::ScpMessage(msg) =>
+						if let ScpStatementPledges::ScpStExternalize(stmt) = &msg.statement.pledges
+						{
+							let txset_hash = get_tx_set_hash(stmt);
+							ov_conn_locked
+								.send(StellarMessage::GetTxSet(txset_hash))
+								.await
+								.unwrap();
+						},
+					StellarMessage::TxSet(set) => {
+						tx_set_vec_clone.lock().await.push(set);
+						ov_conn_locked.disconnect().await.expect("failed to disconnect");
+						break
+					},
+					_ => {},
 				},
 				_ => {},
-			},
-			_ => {},
+			}
 		}
-	}
+	})
+	.await
+	.expect("time has elapsed");
+
+	overlay_connection
+		.lock()
+		.await
+		.disconnect()
+		.await
+		.expect("Should be able to disconnect");
```

Comment:
Do we really need to disconnect twice? There already is a disconnect happening in l. 127, so I'm not sure if/why we need both.